[Bug]: trim_messages does not work, delete tool calls

### What happened?

Umm.. so this bug is still there: https://github.com/BerriAI/litellm/issues/6505

This line removes the tool calls from the messages list, expecting the tool calls to be added back after trimming, but then the size check short-circuits here:

https://github.com/BerriAI/litellm/blob/f66029470f91c3a5452fd963a887544db10c8c5e/litellm/utils.py#L5486

and returns the now truncated list. Then I get errors because my tool call result is missing.

The unit test with tool calls appears to pass coincidentally because it artificially sets `max_tokens=1` so the logic that the token limit isn't exceeded yet is never triggered:

https://github.com/BerriAI/litellm/blob/c494b2cb85e36313730809ce292f739b76275dd1/tests/litellm_utils_tests/test_utils.py#L183


### Relevant log output

```shell

```

### Are you a ML Ops Team?

No

### What LiteLLM version are you on ?

v1.60.0

### Twitter / LinkedIn details

_No response_

You guys know we have Claude now, right? 😂  I guess the point is that you know that **I have** Claude (who gets really excited when it sees the bug, haha, but I had to hint at it):

<img width="727" alt="Image" src="https://github.com/user-attachments/assets/23e2aa65-d182-42d1-82dd-3ded98dadd7a" />


```
import pytest
from unittest.mock import patch
import copy
from litellm.utils import trim_messages

# Test fixtures
@pytest.fixture
def sample_messages():
    return [
        {"role": "user", "content": "Hello"},
        {"role": "assistant", "content": "Hi there"},
        {"role": "tool", "content": "Tool response 1"},
        {"role": "tool", "content": "Tool response 2"}
    ]

@pytest.fixture
def mock_token_counter():
    with patch('litellm.token_counter', return_value=50) as mock:
        yield mock

def test_tool_messages_preserved_when_under_limit(sample_messages, mock_token_counter):
    """
    Test that tool messages are preserved when messages are under the token limit.
    This test verifies the bug where tool messages are dropped even when no trimming is needed.
    """
    # Call the trim_messages function
    result = trim_messages(
        messages=copy.deepcopy(sample_messages),
        max_tokens=100  # Set high enough that no trimming should occur
    )
    
    # Verify total message count
    assert len(result) == len(sample_messages), \
        f"Expected {len(sample_messages)} messages, but got {len(result)}. Tool messages were lost!"
    
    # Get tool messages
    original_tool_messages = [m for m in sample_messages if m["role"] == "tool"]
    result_tool_messages = [m for m in result if m["role"] == "tool"]
    
    # Verify tool message count
    assert len(result_tool_messages) == len(original_tool_messages), \
        "Tool messages were dropped even though no trimming was needed!"
    
    # Verify tool message content
    assert result_tool_messages == original_tool_messages, \
        "Content of tool messages was modified!"

def test_messages_order_preserved(sample_messages, mock_token_counter):
    """
    Test that the order of messages is preserved when no trimming is needed.
    This ensures that tool messages stay at the end and in their original order.
    """
    result = trim_messages(
        messages=copy.deepcopy(sample_messages),
        max_tokens=100
    )
    
    assert result == sample_messages, \
        "Message order was modified when no trimming was needed!"

def test_zero_tool_messages(mock_token_counter):
    """
    Test that the function works correctly with no tool messages present.
    This ensures the tool message handling code doesn't affect regular messages.
    """
    messages = [
        {"role": "user", "content": "Hello"},
        {"role": "assistant", "content": "Hi there"}
    ]
    
    result = trim_messages(
        messages=copy.deepcopy(messages),
        max_tokens=100
    )
    
    assert result == messages, \
        "Regular messages were modified when no tool messages present!"
```

```
FAILED tests/test_trim_messages.py::test_tool_messages_preserved_when_under_limit - AssertionError: Expected 4 messages, but got 2. Tool messages were lost!
FAILED tests/test_trim_messages.py::test_messages_order_preserved - AssertionError: Message order was modified when no trimming was needed!
```

--- a/litellm/utils.py
+++ b/litellm/utils.py
@@ -5829,6 +5829,7 @@ def trim_messages(
     """
     # Initialize max_tokens
     # if users pass in max tokens, trim to this amount
+    original_messages = messages
     messages = copy.deepcopy(messages)
     try:
         if max_tokens is None:
@@ -5912,7 +5913,7 @@ def trim_messages(
         verbose_logger.exception(
             "Got exception while token trimming - {}".format(str(e))
         )
-        return messages
+        return original_messages
 
 
 from litellm.caching.in_memory_cache import InMemoryCache

--- a/litellm/utils.py
+++ b/litellm/utils.py
@@ -5867,7 +5867,7 @@ def trim_messages(
 
         # Do nothing if current tokens under messages
         if current_tokens < max_tokens:
-            return messages
+            return messages + tool_messages
 
         #### Trimming messages if current_tokens > max_tokens
         print_verbose(

--- a/litellm/utils.py
+++ b/litellm/utils.py
@@ -5857,6 +5857,7 @@ def trim_messages(
             if message["role"] != "tool":
                 break
             tool_messages.append(message)
+        tool_messages.reverse()
         # # Remove the collected tool messages from the original list
         if len(tool_messages):
             messages = messages[: -len(tool_messages)]

--- a/tests/litellm_utils_tests/test_utils.py
+++ b/tests/litellm_utils_tests/test_utils.py
@@ -248,6 +248,10 @@ def test_trimming_with_tool_calls():
     assert len(result) == num_tool_calls
     assert result == messages[-num_tool_calls:]
 
+    result = trim_messages(messages=messages, max_tokens=999)
+    # message length is below max_tokens, so output should match input
+    assert messages == result
+
 
 def test_trimming_should_not_change_original_messages():
     messages = [

--- a/tests/litellm_utils_tests/test_utils.py
+++ b/tests/litellm_utils_tests/test_utils.py
@@ -238,11 +238,15 @@ def test_trimming_with_tool_calls():
             "content": '{"location": "Paris", "temperature": "22", "unit": "celsius"}',
         },
     ]
-    result = trim_messages(messages=messages, max_tokens=1, return_response_tokens=True)
+    num_tool_calls = 3
+
+    result = trim_messages(messages=messages, max_tokens=1)
 
     print(result)
 
-    assert len(result[0]) == 3  # final 3 messages are tool calls
+    # only trailing tool calls are returned
+    assert len(result) == num_tool_calls
+    assert result == messages[-num_tool_calls:]
 
 
 def test_trimming_should_not_change_original_messages():

--- a/tests/litellm_utils_tests/test_utils.py
+++ b/tests/litellm_utils_tests/test_utils.py
@@ -1,4 +1,5 @@
 import copy
+import logging
 import sys
 import time
 from datetime import datetime
@@ -280,6 +281,53 @@ def test_trimming_with_model_cost_max_input_tokens(model):
         get_token_count(trimmed_messages, model=model)
         < litellm.model_cost[model]["max_input_tokens"]
     )
+
+
+def test_trimming_with_untokenizable_field(caplog: pytest.LogCaptureFixture) -> None:
+    from litellm.types.utils import ChatCompletionMessageToolCall, Function, Message
+
+    messages = [
+        {
+            "role": "system",
+            "content": "You are a helpful assistant.",
+        },
+        {
+            "role": "user",
+            "content": "What's the weather like in San Francisco?",
+            # non-string values will cause the tokenizer to raise an exception
+            "user_id": 123,
+        },
+        Message(
+            content=None,
+            role="assistant",
+            tool_calls=[
+                ChatCompletionMessageToolCall(
+                    function=Function(
+                        arguments='{"location": "San Francisco, CA", "unit": "celsius"}',
+                        name="get_current_weather",
+                    ),
+                    id="call_G11shFcS024xEKjiAOSt6Tc9",
+                    type="function",
+                ),
+            ],
+            function_call=None,
+        ),
+        {
+            "tool_call_id": "call_G11shFcS024xEKjiAOSt6Tc9",
+            "role": "tool",
+            "name": "get_current_weather",
+            "content": '{"location": "San Francisco", "temperature": "72", "unit": "fahrenheit"}',
+        },
+    ]
+
+    # trim_messages() catches the exception raised by the tokenizer and logs an error
+    with caplog.at_level(level=logging.ERROR, logger="LiteLLM"):
+        trimmed_messages = trim_messages(messages, max_tokens=999)
+
+    assert trimmed_messages == messages
+
+    assert len(caplog.records) >= 1
+    assert "Got exception while token trimming" in caplog.text
 
 
 def test_aget_valid_models():

